etc/docker/kernel-image-puller/kernel_image_puller.py

```diff
@@ -0,0 +1,124 @@
+import os
+import logging
+import time
+import queue
+import requests
+from threading import Thread
+from docker.client import DockerClient
+from docker.errors import NotFound, APIError
+
+gateway_host = os.getenv("KIP_GATEWAY_HOST", "http://localhost:8888")
+num_pullers = int(os.getenv("KIP_NUM_PULLERS", "2"))
+num_retries = int(os.getenv("KIP_NUM_RETRIES", "3"))
+interval = int(os.getenv("KIP_INTERVAL", "300"))
+log_level = os.getenv("KIP_LOG_LEVEL", "INFO")
+
+docker_client = DockerClient.from_env()
+
+logging.basicConfig(format='[%(levelname)1.1s %(asctime)s %(name)s.%(threadName)s] %(message)s')
+
+
+def get_kernelspecs():
+    """Fetches the set of kernelspecs from the gateway, returning a dict of configured kernel specs"""
+    end_point = '{}/api/kernelspecs'.format(gateway_host)
+    logger.info("Fetching kernelspecs from '{}' ...".format(end_point))
+    response = requests.get(end_point)
+    return response.json()
```

Comment:
When server is unhealthy, this would raise `ValueError`. To clarify response error, I think it is better to raise clear error such as `HTTPError`
```
if not r.ok:
    raise requests.exceptions.HTTPError('Server reponse: {}'.format(r.status_code))
```